sklearn/tests/test_base.py

```diff
@@ -252,3 +258,39 @@ def test_score_sample_weight():
                                    sample_weight=sample_weight),
                          msg="Unweighted and weighted scores "
                              "are unexpectedly equal")
+
+
+def test_pickle_version_warning():
+    # check that warnings are raised when unpickling in a different version
+
+    # first, check no warning when in the same version:
+    iris = datasets.load_iris()
+    tree = DecisionTreeClassifier().fit(iris.data, iris.target)
+    tree_pickle = pickle.dumps(tree)
+    assert_no_warnings(pickle.loads, tree_pickle)
+
+    # check that warning is raised on different version
+    tree_pickle_other = tree_pickle.replace(sklearn.__version__.encode(),
+                                            b"something")
+    message = ("Trying to unpickle estimator DecisionTreeClassifier from version "
+               "{0} when using version {1}. This might lead to breaking "
+               "code or invalid results. Use at your own risk.".format(
+                   "something", sklearn.__version__))
+    assert_warns_message(UserWarning, message, pickle.loads, tree_pickle_other)
+
+    # check that not including any version also works:
+    # don't do this at home, kids
+    from sklearn.base import BaseEstimator
+    old_getstate = BaseEstimator.__getstate__
+    del BaseEstimator.__getstate__
+    # tree lost its getstate, like pre-0.18
+    assert_false(hasattr(tree, "__getstate__"))
+    tree_pickle_noversion = pickle.dumps(tree)
+    message = message.replace("something", "pre-0.18")
+    # check we got the warning about using pre-0.18 pickle
+    assert_warns_message(UserWarning, message, pickle.loads, tree_pickle_noversion)
+    BaseEstimator.__getstate__ = old_getstate
+
+    # check that no warning is raised for external estimators
+    DecisionTreeClassifier.__module__ = "notsklearn"
+    assert_no_warnings(pickle.loads, tree_pickle_noversion)
```

Comment:
You probably need to reset the `__module__` after the `assert_no_warnings`. I am guessing this is the reason of the test failures.